### ISSUE TITLE
Cache vs. @:generic method

### ISSUE THREAD

**Issue by Simn** [HaxeFoundation/haxe#10635]
```haxe
// Something.hx
class Something {
	@:generic
	static public function append<T>(a:Array<T>, b:Array<T>) {
		for (e in b) {
			a.push(e);
		}
		return a;
	}
}
```

```haxe
// Main.hx
class C {
	public function new() {}
}

function main() {
	Something.append([new C()], [new C()]);
}
```

Triggering diagnostics twice reports this:

```
source/Main.hx:6: characters 2-40 : Cannot create field append_C due to type mismatch
source/Something.hx:3: lines 3-8 : ... Conflicting field was defined here
source/Main.hx:6: characters 2-40 : error: C should be C
source/Main.hx:6: characters 2-40 : ... have: (...) -> Array<C>
source/Main.hx:6: characters 2-40 : ... want: (...) -> Array<C>
```

This is probably not about diagnostics itself, but rather about some missing dependency.

**Comment by Simn** [comment]
We have to consider `@:genericClassPerMethod` here, which I totally forgot about.

**Comment by Simn** [comment]
There's another problem (which was failing before my fixes too). To reproduce:

```haxe
// M.hx
class M {
	@:generic static public function test<T>(t:T) {}
}
```

```haxe
// Main.hx
class Main {
	static function main() {
		trace(M);
	}
}
```

Compile that so that it is cached, then change Main.hx to this:

```haxe
class Main {
	static function main() {
		trace(M);
		M.test(Main);
	}
}
```

Compiling this works fine, but when compiling _again_ we get run-time errors because the expanded generic method is missing.

I think this must be `cl_restore` resetting `M` to its original state. During the second compilation, `M` is reused because when we first see it at the `trace(M)`, there's no reason not to reuse it. The `M.test(Main)` call then expands the generic method, modifying the `tclass` that is cached. But on the next compilation, we end up reusing `M` again because neither it nor `Main` have changed. This then calls `cl_restore`, which isn't aware of the new field.

The example might seem artificial, but this is the kind of problem we'll have to consider for anything that modifies existing cache state. I suppose we can fix this by marking cached modules which are subject to `add_dependency` and then make sure we call `save_class_state` on them too.

**Comment by Simn** [comment]
Actually this is pretty dicey because we want to run the entire post-processing up to `save_class_state` on such newly added fields, but we _don't_ want to run it again on already processed ones... This might require some notion of field-level cache tracking.

**Comment by Simn** [comment]
This has been resolved (again) by #10643, but [its successor](#10653) has already been found.
